packages/styles/src/use-style-injection/useStyleInjection.ts

```diff
@@ -0,0 +1,75 @@
+import * as React from "react";
+import { useInsertionPoint } from "./InsertionPointProvider";
+
+/* eslint-disable @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-assignment -- Workaround for https://github.com/webpack/webpack/issues/14814 */
+const maybeUseInsertionEffect: typeof React.useLayoutEffect =
+  (React as any)[`${"useInsertionEffect"}${""}`] ?? React.useLayoutEffect;
+/* eslint-enable @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-assignment */
+
+type WindowLike = Window & typeof globalThis;
+
+export interface UseComponentCssInjection {
+  testId?: string;
+  css: string;
+  window?: WindowLike;
+}
+
+type StyleElementMap = Map<
+  string,
+  { styleElement: HTMLStyleElement | null; count: number }
+>;
+
+const windowSheetsMap = new WeakMap<WindowLike, StyleElementMap>();
```

Comment:
I think we need some inline comment about what we expect from this object, when and how it will change, to cover the "algorithm" of the hook below